packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -0,0 +1,328 @@
+        } else {
+          assertKeyDoesNotExist(key);
+          backingMap.set(key, value);
+        }
+      },
+      get(key) {
+        const vkey = virtualObjectKey(key);
+        if (vkey) {
+          const rawValue = syscall.vatstoreGet(vkey);
+          assert(rawValue, details`${q(keyName)} not found: ${key}`);
+          return m.unserialize(JSON.parse(rawValue));
+        } else {
+          assertKeyExists(key);
+          return backingMap.get(key);
+        }
+      },
+      set(key, value) {
+        const vkey = virtualObjectKey(key);
+        if (vkey) {
+          assert(
+            syscall.vatstoreGet(vkey),
+            details`${q(keyName)} not found: ${key}`,
+          );
+          syscall.vatstoreSet(vkey, JSON.stringify(m.serialize(harden(value))));
+        } else {
+          assertKeyExists(vkey);
+          backingMap.set(key, value);
+        }
+      },
+    });
+  }
+
+  /**
+   * Make a new kind of virtual object.
+   *
+   * @param {*} instanceMaker  A function of the form `instanceMaker(state)` that
+   *    will return a representative instance wrapped around the given state.
+   *
+   * @returns {*} a maker function that can be called to manufacture new
+   *    instance of this kind of object.  The parameters of the maker function
+   *    are those of the `initialize` method implemented in the representative
+   *    produced by the `instanceMaker` parameter.
+   */
+  function makeKind(instanceMaker) {
+    const kindID = `${allocateExportID()}`;
+    let nextInstanceID = 1;
+
+    function makeRepresentative(innerSelf) {
+      function ensureState() {
+        if (!innerSelf.rawData) {
+          innerSelf = cache.lookup(innerSelf.instanceKey);
+        }
+      }
+
+      function wrapData() {
+        const activeData = {};
+        for (const prop of Object.getOwnPropertyNames(innerSelf.rawData)) {
+          Object.defineProperty(activeData, prop, {
+            get: () => {
+              ensureState();
+              return m.unserialize(innerSelf.rawData[prop]);
+            },
+            set: value => {
+              ensureState();
+              innerSelf.rawData[prop] = m.serialize(value);
+            },
+          });
+        }
+        return harden(activeData);
+      }
+
+      const representative = instanceMaker(wrapData());
+      delete representative.initialize;
+      harden(representative);
+      cache.remember(innerSelf);
+      valToSlotTable.set(representative, innerSelf.instanceKey);
+      return representative;
+    }
+
+    function reanimate(instanceKey) {
+      return makeRepresentative(cache.lookup(instanceKey));
+    }
+    kindTable.set(kindID, reanimate);
+
+    function makeNewInstance(...args) {
+      const instanceKey = `o+${kindID}/${nextInstanceID}`;
+      nextInstanceID += 1;
+
+      const initializationData = {};
+      const tempInstance = instanceMaker(initializationData);
+      tempInstance.initialize(...args);
```

Comment:
I added #1938 to explore this further

Reply:
Hmmm.  This one is tricky, since `initialize` is called before the state has been wrapped (since the wrapping needs the result of `initialize` to act upon).  If `initialize` passes the instance to somebody else, they might call one of the other methods, which would then operate on the unwrapped state directly.  This would be OK inside the call stack beneath `initialize`, since the net result would basically just be a more complicated initialization.  However, if one of _those_ callees were to squirrel the "instance" away somewhere and invoke it later, it would be operating on a disconnected, non-persistent, non-consistent private copy of the state, which would be bad and wrong.  I think this means that the wrapping operation needs to modify the state object in place (which you previously talked me into not doing -- in fact, now that I think about it this may have been the reason why I originally wrote the code to work that way).

In any case, some kind of verification that this works OK needs to be added to the unit tests.

Reply:
In writing up #1938, I concluded our Purse example won't work work at all until we add the `tempInstance` to `valToSlot` before calling `initialize`. We should add that Purse example to a unit test, and either include that in this PR (necessitating the `valToSlot` addition in the PR too), or shortly after landing (precluding this feature's use in ERTP until we get #1938 fixed).

Reply:
>  I think this means that the wrapping operation needs to modify the state object in place (which you previously talked me into not doing -- in fact, now that I think about it this may have been the reason why I originally wrote the code to work that way).

Eeeeinteresting... and even more subtle than what I was investigating. Yeah, sorry if I undid that. Definitely needs some tests either way.

Reply:
Latest version copes with having the representative being handed to somebody by `initialize`.  I believe the fix addresses the issues you describe in #1938 as well, but I'm working on a unit test to verify this for sure.

Reply:
good, thanks. I'm still nervous about this case, the full test would be if that "somebody" uses the representative as the value in a weakstore (serializing it into just the ID), and *then* reads that value back out again (creating a new representative), and then calls methods on the new representative.

And I strongly suspect that the presence of any code after `initialize()` returns means that the first and second representatives are not functionally equivalent. Or at least I *am* sure that the lack of any code after that point means they *are* equivalent, which may or may not imply the contrapositive.